### DISCUSSION TITLE
KeeWeb Favicon Grabber Service

### DISCUSSION THREAD

**Discussion by Aetherinox** [keeweb/keeweb#2169] [Announcements]
## Summary

Our Favicon Grabber service is now operational again and should work for existing builds of KeeWeb.

I have tested it for the web platform, Windows, and Linux, however, I do not have access to MacOS, but I anticipate that it should operate normally, as none of the back-end code has changed, just the service itself. If you do end up having issues, let me know by submitting an [issue](https://github.com/keeweb/keeweb/issues).

<div align="center">

![eulXdQYHBF](https://github.com/keeweb/keeweb/assets/118329232/bbf1eb4e-5d59-44e3-80a9-4ec6740e0491)

</div>

## Important Notes

There are request limits in place, so please ensure you are not sending multiple requests per second. If you are grabbing icons for each of your entries in KeeWeb normally; you should not hit these limits. There is also a daily limit, but this has been set generously; unless you have a very large vault with many hundreds of websites that you need to fetch icons for in a single day; this should be no issue as well. If you try to fetch a favicon and it doesn't respond; wait a minute or so, and try again.

We have also decided to create a KeeWeb favicon repo available at: https://github.com/keeweb/favicon-cdn. This has been implemented for several reasons.

1. In the instance that a website does not have a favicon returned; we can override it by adding our own. The favicon grabber will search the cdn repo for an icon and grab that with priority over the actual favicon the website serves. Surprisingly, there are still quite a few websites out there without a favicon. Or have not specified a fav icon for subdomains. This repo allows for us to define favicons for users who may not want to manually upload their own.

2. Same favicons may not look great. Some of them are older, don't offer transparency, or are hard to see due to them only having the smaller 16 x 16 available. This repo allows us to implement higher-quality icons.

The repo will not house every favicon known to man and is only for overrides. If a website has a high quality favicon offered, then it shouldn't need to be in our repo.

Contributions are welcome, but please ensure you read that repo's [README](https://github.com/keeweb/favicon-cdn/blob/master/README.md) before adding submissions.

Some users have asked if we can implement a way for users to specify their own favicon service, and we are currently looking into addressing this request. There are some things to take into consideration such as CORS, but we've been testing a few possibilities, and one configuration looks promising, but it is not ready for release.

<br />

---

<br />

## Support for Self-Hosters
Much like myself, I am sure there are other people out there who enjoy self-hosting a lot of their applications, including apps like Syncthing, Duplicati, Duplicacy, etc.

Today I've added support for self-hosted applications

![t4pFs2k6jI](https://github.com/keeweb/keeweb/assets/118329232/547d749b-a76d-42cf-8381-f35895e8e244)

This means that if you have a locally hosted webapp such as Syncthing, you can grab the icon for that service by using the private ip to the service.
- `127.0.0.1:PORT`
- `192.168.x.x:PORT`
- `172.16.x.x:PORT`

For users who wish to contribute different icons for self-hosted services; all private ips are seen internally as `127.0.0.1` just to keep things clean within the repo. All icons should be placed within: https://github.com/keeweb/keeweb-favicon-cdn/tree/master/1

If you change the default port to webapps such as Syncthing from something other than port `8384`, it will not fetch the icon. It would be impossible and time consuming to start pinging all of those services, so we'll stick to default ports for now.

For a list of supported web-apps, view https://github.com/keeweb/favicon-cdn?tab=readme-ov-file#self-hosted-ports

**Comment by GamerClassN7** [comment]
Hello, it is nice to see that something is happening around KeeWeb (i absolutely love the project).
Would be possible to implement setting option to set favicon scraper as default action for saving new entry, or updating existing entry which previously had favicon  ?

Thank you in advance for any response!

**Reply by Aetherinox** [reply]
Are you referring to it automatically fetching a favicon when you create a new entry instead of manually having to click the download button?

If so, yes, it's possible to add an option to automatically fetch.

If memory serves me right, KeePassXC has a button to fetch the favicon which is to the right of the web address, and it also has a "Download All Favicons".

Create a new [Issue](https://github.com/keeweb/keeweb/issues) outlining the feature as a feature request, and I'll add it into the pipeline. I have a lot of things with this project going on at the moment, so I can forget in five minutes.
